### ISSUE TITLE
 Custom metric/callback issues [feature request?]

### ISSUE THREAD

**Issue by roya0045** [keras-team/keras#8269]
Greetings,

I've been trying to create a full logger to catch all the data going through the model during training.

I am aware that the callbacks only take in the  "logs" which seem to be the results of the metrics.

So I tried to create a custom metric to compute the error and log the data but it seems that something doesn't work. I am confused about how metrics are handled at runtime since the print statement that I included in my metric ( and any changes to instance or global variable) only happens before training once and doesn't seem to happen again. 

Would there be a way to properly get the full logs with a callback or a metric? (either by tweaking my prototype or by implementing a new feature if such a thing can be added.)

My current prototype is [the following.](https://gist.github.com/roya0045/94817bf98449111a002c1c05b3a83862)

Thanks

**Comment by roya0045** [comment]
I'm guessing the issue might be coming from the metric weighter ([line 430 of trainning.py ](https://github.com/fchollet/keras/blob/3b15ee6954b688b9490fd885d628939fd35f76ca/keras/engine/training.py#L430) ), but I may be off the mark on that one. This line doesn't explain why the custom metric is only called once during the model building.

**Comment by Dref360** [comment]
Your print statement is only printed once because this is happening at the graph-definition step. I would recommend that you get more familiar with your backend compilation phase.

Each op defines a node in a graph called a computation graph. At runtime, only the graph is run so your methods are only called once to define the graph. If you want to do logging, you need a logging operation inside your graph like tf.Print. 

Please close as this is not an issue.

**Comment by roya0045** [comment]
Issue tracker may be used to ask for a new feature, considering that there is a already a logger that provides average metrics at each evaluation step, I think that a more precise logger could be a useful improvement, implementing advanced login could be a useful feature. The only alternative currently would be to use a batch size of 1 to do any logging.

Also even when implementing tf.print or the backend print_tensor in the custom metric, but it doesn't print anything, and even if it did those printed values would need to be caught and send to a file for logging, ( this could be triggered with a callback). But I still don't see how to catch the tensor of every predicted values at each training step. 

But in the meantime I'll look into the details of the compilation phase to try to find anything that could help.

Thanks for the response @Dref360

**Comment by w121211** [comment]
I have the same problem, `K.print_tensor()` seems not working in the loss function, and leads me no where to check the value of `y_pred` on each iteration during training.
Hope there function will be added.

**Comment by roya0045** [comment]
@w121211 For my use, the workaround that I found was to train the whole model, and once that's done to simply use the prediction function on each data, gather those points and compare them with the true value to document the error of my trained model. There is no way to check the predicted value during training but it is highly possible that those values will never happen with the trained model anyway due to the parameter updates.
